flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java

```diff
@@ -1515,8 +1516,22 @@ public void reportPayload(ResourceID resourceID, Void payload) {
 		}
 
 		@Override
-		public CompletableFuture<Void> retrievePayload(ResourceID resourceID) {
-			return CompletableFuture.completedFuture(null);
+		public CompletableFuture<AccumulatorReport> retrievePayload(ResourceID resourceID) {
```

Comment:
Let's add `validateRunsInMainThread` as a first statement. That way we enforce that this method really runs in the main thread context.